README
pod2text Device::DSE::Q15734::Plotter.pm > README

If this is still here it means the programmer was too lazy to create the readme file.

You can create it now by using the command shown above from this directory.

At the very least you should be able to use this set of instructions
to install the module...

perl Makefile.PL
make
make test
make install

If you are on a windows box you should use 'nmake' rather than 'make'.


Changes
Revision history for Perl module Device::DSE::Q15734::Plotter

0.01 Mon May  3 18:27:15 2010
    - original version; created by ExtUtils::ModuleMaker 0.51




Makefile.PL


use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
my $cgi = prompt("Where would you like kilngraph.pl & kilnserve.pl installed ", 
						"/var/www/cgi-bin");

WriteMakefile(
    NAME         => 'Device::Kiln',
    VERSION_FROM => 'lib/Device/Kiln.pm', # finds \$VERSION
    AUTHOR       => 'David Peters (davidp@electronf.com)',
    ABSTRACT     => 'Plot Graph of values from Q1573 over time.',
    EXE_FILES	 => ['lib/Device/kilngraph.pl',
    				'lib/Device/kilnserver.pl'],
    INSTALLSITESCRIPT => $cgi,
    PREREQ_PM    => {
                     'Test::Simple' => 0.44,
                     'SVG' => 0.01,
                     'SVG::TT::Graph::TimeSeries' => 0,
                     'HTTP::Date' => 0
                    },
);


lib/Device/kilngraph.pl
#!/usr/bin/perl
use Device::Kiln;
use Device::Kiln::Orton;
use strict;
use CGI;      # or any other CGI:: form handler/decoder
use CGI::Ajax;
use Template;
use Data::Dumper;

  my %action_table = (
    graph_rm => \&graph_rm,
    main_rm => \&main_rm,
    test_rm => \&test
  );



my $meter = Device::Kiln->new({
		width => 1024,
		height => 800,
	});





my $cgi = new CGI;
my $rm = $cgi->param("rm") || "main";
$rm .= "_rm";

&{ $action_table{$rm} }($cgi,$meter);


sub graph_rm ($$){	

	my ($cgi,$meter) = @_;
	
	my $cone = $cgi->param("cone");
	
	print $cgi->header({-type=>'image/svg+xml'});
#	print $meter->graph( {
#			cone => $cone,
#			warmuptime => $cgi->param("warmuptime"),
#			warmuptemp => $cgi->param("warmuptemp"),
#			warmupramp => $cgi->param("warmupramp"),
#			fullgraph => 0,
#		});

	print $meter->graph({$cgi->Vars})
	
}	
		
sub main_rm  ($$) {
	
	my ($cgi,$meter) = @_;
	
	my $pjx = new CGI::Ajax( 'exported_func' => \&update );
	print $pjx->build_html( $cgi, \&Show_HTML);
}

sub update {
	
}

sub test ($$){
	
	my ($cgi,$meter) = @_;
	
	print $meter->{test} . "\n";
	
}

sub debug {
	
	my $msg = shift;
	
	open( my $dfh, ">>", "/tmp/kilnserver.dbg");
	print $dfh (scalar localtime) . " : " . $msg . "\n";
	close($dfh);
	
}

#<object data='/cgi-bin/kilngraph.pl?rm=graph' id='kilngraph' type='image/svg+xml' width=1024 height=600 >

sub Show_HTML {
    my $html = "
    <HTML>
	<head>
	<script type='text/javascript'>
		
		
	function newImage(){
		if(document.images){
			var cone = params.cone.value;
			var conerate = params.conerate.value;
			var warmuptime = params.warmuptime.value;
			var warmuptemp = params.warmuptemp.value;
			var warmupramp = params.warmupramp.value;
			var fireuprate = params.fireuprate.value;
			document.getElementById('kilngraph').src = '/cgi-bin/kilngraph.pl?rm=graph&'
				 + 'cone=' + cone + '&'
				 + 'conerate=' + conerate + '&'
				 + 'warmuptime=' + warmuptime + '&'
				 + 'warmuptemp=' + warmuptemp + '&'
				 + 'warmupramp=' + warmupramp + '&'
				 + 'fireuprate=' + fireuprate + '&'
				 + Date.parse(new Date().toString());
		
		} 
		
	}
	
	function cycle() {
		newImage();
		setTimeout(cycle,15000);
	}
	
	
	</script>
	</head>
	<body onload='cycle()'>
	<div id='resultdiv'></div>
	<form name='params' >
	<TABLE BORDER=1>
	<TR>
	<TD>
	Warm Up Temp <select name=warmuptemp onchange='newImage()'>
	<option value=100>100
	<option value=150>150
	<option value=200>200
	</select>
	</TD>
	<TD>
	Warm Up Ramp <select name=warmupramp onchange='newImage()'>
	<option value=100>100
	<option value=150>150
	<option value=200>200
	<option value=250>250
	<option value=300>300
	</select> C/Hour
	</TD>
	<TD>
	Warm Up Time <select name=warmuptime onchange='newImage()'>
	<option value=30>30
	<option value=60>60
	<option value=90>90
	<option value=120>120
	</select> Minutes
	</TD>
	</TR>
	<TR>
	<TD>
	Fire Up Rate <select name=fireuprate onchange='newImage()'>
	<option value=150>150
	<option value=200>200
	<option value=300>300
	<option value=400>400
	</select> C/Hour
	</TD>
	<TD>
	Last 100C(Degrees C/hour) <select name=conerate onchange='newImage()'>
	<option value=15>15</option>
	<option value=60>60</option>
	<option value=150>150</option>
	</select>
	</TD>
	<TD>
	Cone 	<select name=cone onchange='newImage()'>";
	
	my $cones = Device::Kiln::Orton->hashref();
		foreach my $coneconfig ( 
			sort { $cones->{$a}->{seqnum} <=> $cones->{$b}->{seqnum} }  keys %$cones 
	) {
		
		$html .= "<option value = '$coneconfig'>$coneconfig\n";
	}
	
	$html .= "
	</select>
	</TD>
	</TR>
	</TABLE>
	</form>


	<BR>

<TABLE BORDER=1>
<TR>
<TD>
	<iframe id='kilngraph' width=1024 height=600 >	</iframe>
</TD>
<TD VALIGN=TOP>
	<TABLE VALIGN='top'>
	<TR>
	<TD>
	<B>Time</B>
	</TD>
	<TD>
	<B>Temp</B>
	</TD>
	</TR>
	<TR>
	<TD>
	</TD>
	<TD>
	</TD>
	</TR>
	</TABLE
</TD>
</TR>
</TABLE>
</HTML>
 	";
    return $html;
}

lib/Device/Kiln.pm
package Device::Kiln;

use Device::DSE::Q1573;
use Device::Kiln::Orton;
use SVG::TT::Graph::TimeSeries;
use HTTP::Date qw(time2iso str2time);
use Data::Dumper;

use strict;

BEGIN {
	require Device::DSE::Q1573;
	use Exporter ();
	use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
	$VERSION     = '0.02';
	@ISA         = qw(Exporter);
	@EXPORT      = qw();
	@EXPORT_OK   = qw();
	%EXPORT_TAGS = ();

}

sub new {
	
	my $class = shift;
	my %config = %{$_[0]};
	


	my $self = bless( {}, ref($class) || $class );
	
	
	if( defined $config{'serialport'} ) {
		$self->{meter}      = Device::DSE::Q1573->new($config{serialport});
		$self->{serialport} = $config{serialport};
	}
		

	$self->{width}      = $config{'width'};
	$self->{height}     = $config{'height'};
	$self->{interval}   = $config{'interval'};
	$self->{rampuprate} = 150;
	$self->{tid}        = undef;
	$self->{datafile}   = '/tmp/q1573.tmp';
	$self->{cones}		= {
			'1 - 018' =>  	{ 
							60  => 712,
							100 => 722,
							150 => 732
							},
							
			'2 - 017' =>  	{
							60  => 736,
							100 => 748,
							150 => 761
							},
							
			'3 - 016' =>	{
							60  => 769,
							100 => 782,
							150 => 794
							},
							
						};
				
	$self->{maxtemp}	= [""];
	return $self;
}

sub run {


	my $self = shift;

	print $self->{file} . "\n";
	my $time  = 0;
	my $count = 0;

	my $fh;

	open( $fh, ">>", $self->{datafile} );
	$fh->autoflush(1);

	sleep 1;

	while (1) {
		
		my $time = time();
		my $raw = $self->{meter}->rawread();
		#print $raw. "\n";
		$self->{setting} = substr( $raw, 0,  2 );
		$self->{value}   = substr( $raw, 2,  7 ) + 0;
		$self->{units}   = substr( $raw, 11, 1 );

		$self->{value} =~ s/ //g;

		#print $self->{value} . " ";
		print time2iso() . "\n";

		print $fh time2iso() . "|" . $self->{value}  . "\n";
		
		
		while( time() < ($self->{interval} + $time) ){sleep 1;};
		print time() . " : $time\n";
		

	}

}

sub graph {

	my $self = shift;
	my $config = shift;
		
	my @data;
	my @ideal;
	my $fh;
	my ($startperiod, $startvalue);
	
	
	$config->{conemax} = Device::Kiln::Orton->hashref()->{$config->{cone}}->{$config->{conerate}};
	#$config->{conemax} = $self->{cones}->{$config->{cone}}->{$config->{conerate}};
	

	#@data = ( [ time2iso(), $self->{value} ] );
	
	open( $fh, "<", $self->{datafile} );
	my $count = 0;
	while(<$fh>) {


		chomp;
		push @data, [split /\|/, $_];
		
		
		if($count == 0 ) {
			$startperiod = $data[0][0];
			$startvalue = $data[0][1];
			$count++;
		}
		
				
	}

	debug("-------------------------");
	
	debug("Warmupramp :" . $config->{warmupramp});
	debug("Warmuptemp :" . $config->{warmuptemp});
	debug("Warmuptime :" . $config->{warmuptime});
	debug("Fireuprate :" . $config->{fireuprate});
	debug("Conerate   :" . $config->{conerate});
	debug("Cone       :" . $config->{cone});
	debug("Cone Max   :" . $config->{conemax});
	
	
	
	my $endperiod = @{@data[@data-1]}[0];
	my $duration = str2time($endperiod) - str2time($startperiod);
	
	my $value=$startvalue;

	debug("Duration " . $duration);
	
	push @ideal, [$startperiod,$value];
	debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
	

	
	push @ideal, warmupramp($startvalue,$startperiod,$endperiod,$config);

	
	# Warm Up Time
	my $warmupsecs = $config->{warmuptime} * 60;
	
	if($duration > $config->{used} || $config->{fullgraph} == 0) {
		
		push @ideal, warmuptime($startvalue,$startperiod,$endperiod,$config);
		
		# Fire up to Pre Cone Fire	
		if( $duration > $config->{warmupramptime} + $warmupsecs && $config->{fullgraph} == 0) {
			push @ideal, fireuptime($startvalue,$startperiod,$endperiod,$config);
		
			if($duration > $config->{fireuptime}) {
				push @ideal, conefire($startvalue,$startperiod,$endperiod,$config);
			}
		}
	
	}
	
	
	
	my $graph = SVG::TT::Graph::TimeSeries->new(
		{

			# Optional - defaults shown
			'height' => 600,
			'width'  => 1024,

			'x_label_format' => '%H:%M:%S',
			'area_fill'      => 1,

			'max_time_span' => 0,
			'timescale_divisions' => '15 minutes',

		    'rollover_values'   => 1,
		    'show_data_points'  => 1,
		    'show_data_values'  => 1,

			# Stylesheet defaults
			'style_sheet'   => '/graph.css',    # internal stylesheet
			'random_colors' => 0,
			'compress'      => 0,
			'key'		=> 1,
			
		}
	);

	$graph->add_data(
		{
			'data'  => \@data,
			'title' => 'Temperature',
		}
	);
	
	$graph->add_data(
		{
			'data'  => \@ideal,
			'title' => 'Ideal',
		}
	);
	
	
	debug("-------------------------");

	
	return $graph->burn();
	

}


sub warmupramp {

	my ($startvalue,$startperiod,$endperiod,$config) = @_;
	
	my @ideal;
	my $duration = str2time($endperiod) - str2time($startperiod);
	
	#
	# Warmup Ramp
	#
	debug( "-----> warmupramp ");
	$config->{warmupramptime} = ($config->{warmuptemp}-$startvalue)/($config->{warmupramp})*60*60;
	
	
	if($duration <= $config->{warmupramptime} && $config->{fullgraph} == 0) {
		my $value = ($duration/3600) * $config->{warmupramp} + $startvalue;
		push @ideal, [$endperiod,$value];
		debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
		$config->{used} = $duration;
		
	} else {
		push @ideal, [time2iso($config->{warmupramptime}+str2time($startperiod)),$config->{warmuptemp}];
		$config->{used} = $config->{warmupramptime};
		debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
		
	}
	return @ideal;
}

sub warmuptime {
	
	my ($startvalue,$startperiod,$endperiod,$config) = @_;
	my $warmupsecs = $config->{warmuptime} * 60;
	my $duration = str2time($endperiod) - str2time($startperiod);
	my @ideal;
		
	if( $duration < $config->{used} + $warmupsecs && $config->{fullgraph} == 0) {
		
		push @ideal, [$endperiod,$config->{warmuptemp}];
		$config->{used} = $duration;
		
	} else {

		push @ideal, [time2iso($config->{warmupramptime}+str2time($startperiod)+ $warmupsecs),$config->{warmuptemp}];
		$config->{used} = $config->{used} + $warmupsecs; 
		
	}	

	debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
	return @ideal;
	
}	



sub fireuptime {
	
	my ($startvalue,$startperiod,$endperiod,$config) = @_;
	my $duration = str2time($endperiod) - str2time($startperiod);
	
	my @ideal;
	
	my $endtemp = $config->{conemax} - 100;
	my $risetemp = $endtemp - $config->{warmuptemp};
	
		
	my $endfireuptime = $config->{used} + ($risetemp / $config->{fireuprate}) * 3600;
	
	
	if($duration < $endfireuptime ) {
		my $leftover = $duration - $config->{used};
		my $value = $leftover * ($config->{fireuprate} / 3600) + $config->{warmuptemp};
		push @ideal, [$endperiod, $value];
		$config->{used} = $duration;
		
	} else {
		
		
		push @ideal, [time2iso($endfireuptime + str2time($startperiod)),$endtemp];
		$config->{used} = $endfireuptime;
		$config->{fireuptemp} = $endtemp;
	}
	
	
	debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
	return @ideal;	
	
}


sub conefire {
	
	my ($startvalue,$startperiod,$endperiod,$config) = @_;
	my $duration = str2time($endperiod) - str2time($startperiod);
	my @ideal;
	my $fullconetime = 100 / $config->{conerate} * 3600;
	
	if( $duration < $config->{used} + $fullconetime ) {
	 
		push @ideal, [$endperiod, $config->{fireuptemp} + 100 * (($duration-$config->{used})/$fullconetime) ];
		$config->{used} = $duration;
		
	} else {
		
		
		push @ideal, [time2iso($config->{used} + 3600 + str2time($startperiod)), $config->{conemax}];
		$config->{used} += $fullconetime; 
		debug("in");
	}
	
	debug( "Data : " . $ideal[@ideal-1][0] . ", " . $ideal[@ideal-1][1]);
	return @ideal;
	
}

sub debug {
	
	my $msg = shift;
	
	open( my $dfh, ">>", "/tmp/kilnserver.dbg");
	print $dfh (scalar localtime) . " : " . $msg . "\n";
	close($dfh);
	
}

=head1 NAME

Device::Kiln - Graph kiln firing use Data logged from Device::DSE::Q1573 

=head1 SYNOPSIS

  use Device::Kiln;
  my $meter = Device::Kiln->new("/dev/ttyS0");
  


=head1 DESCRIPTION

Sets up a connection to a DSE Q1573 or Metex ME-22 
Digital Multimeter, and allows then plots data to a png file
at regular intervals. 
=head1 USAGE

=head2 new(serialport)

 Usage     : my $meter=Device::Kiln->new("/dev/ttyS0")
 Purpose   : Opens the meter on the specified serial port
 Returns   : object of type Device::Kiln
 Argument  : serial port
 
=head2 rawread();

 Usage     : my $meter->rawread()
 Purpose   : Returns the raw 14 byte string from the meter.
 
=head1 EXAMPLE

use Device::DSE::Q1573;

my $meter = Device::Kiln->new( "/dev/ttyS0" );

while(1) {
	my $data = $meter->read();
	print $data . "\n";
	sleep(1);
}


=head1 AUTHOR

    David Peters
    CPAN ID: DAVIDP
    davidp@electronf.com
    http://www.electronf.com

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

perl(1).

=cut

#################### main pod documentation end ###################

1;
